### PR TITLE
Issue #131: Added D3 and d3-flame-graph third party libraries.

### LINKED ISSUE #131: Bundle d3 and flamegraph dependencies locally
* [ ] Add clear docs in update

https://docs.moodle.org/dev/Plugin_files#readme_moodle.txt

### PR DESCRIPTION
**Pull request by jaypha**
resolves #131
Created lib directory.
Added D3 library
Added d3-flame-graph library
Created lib/thirdpartylibs.xml
Updated flamegraph.mustache to use new library source.

### CODE REVIEW

File: lib/d3-flame-graph/readme_moodle.txt [hunk 1]
```diff
@@ -0,0 +1,17 @@
+
+Home page for d3-flamegraph is https://github.com/spiermar/d3-flame-graph.
+
+You will need NPM isntalled on your system to perfrom the follwoing procedure.
```
Comment (keevan) [inline lib/d3-flame-graph/readme_moodle.txt:4]: Typo `isntalled` => `installed`

Comment (keevan) [reply lib/d3-flame-graph/readme_moodle.txt:4]: perfrom => perform

Comment (keevan) [reply lib/d3-flame-graph/readme_moodle.txt:4]: follwoing => following

File: lib/d3-flame-graph/readme_moodle.txt [hunk 2]
```diff
@@ -0,0 +1,17 @@
...
+Procedure for installing d3-flame-graph
+* Create a suitable directory.
+* Perform the following commands.
+
+$ git clone https://github.com/spiermar/d3-flame-graph.git
+$ cd d3-flame-graph
+$ npm install
+
+* Copy contents of 'dist' directory to lib/d3-flamegraph in plugin.
+* Copy LICENSE to lib/d3-flamegraph in plugin.
```
Comment (keevan) [inline lib/d3-flame-graph/readme_moodle.txt:15]: I wonder if the whole thing could be commands we can use to do a quick update.
```suggestion
* Copy contents of 'dist' and license directory to lib/d3-flamegraph in plugin.
cp -r somewhere/dist somewhere/LICENSE -t lib/d3-flame-graph
```

Comment (keevan) [reply lib/d3-flame-graph/readme_moodle.txt:15]: Also in your example the target folder has a different name from the one you've provide in this commit

lib/d3-flamegraph => lib/d3-flame-graph

File: lib/d3/readme_moodle.txt [hunk 3]
```diff
@@ -0,0 +1,7 @@
+
+Procedure for including D3 library into Moodle.
+
+1. The download website is https://d3js.org.
+2. Download the latest version in a tgz file via a link given on the above page.
+3. Extract from the tgz file and move the contents of the 'package/dist' directory and LICENSE into lib/d3 of the plugin.
+
```
Comment (keevan) [inline lib/d3/readme_moodle.txt:7]: Similarly to the previous point regarding d3-flame-graph, could you make this into a list of commands and the steps as command comments?

An example URL would be nice to go through the update process as easily as possible without needing to do much manual work.

File: lib/d3/readme_moodle.txt [hunk 4]
```diff
@@ -0,0 +1,7 @@
...
```
Comment (keevan) [inline lib/d3/readme_moodle.txt:1]: Remove the first empty line as I don't think it's needed.

File: lib/d3-flame-graph/readme_moodle.txt [hunk 5]
```diff
@@ -0,0 +1,17 @@
+
```
Comment (keevan) [inline lib/d3-flame-graph/readme_moodle.txt:1]: Remove the first empty line as I don't think it's needed.

**Review by keevan — COMMENTED** [review]
The rest is probably fine but I'll need to re-test once the instructions are updated.

Thanks Jason

File: lib/thirdpartylibs.xml [hunk 6]
```diff
@@ -0,0 +1,18 @@
+<?xml version="1.0"?>
+<libraries>
```
Comment (brendanheywood) [inline lib/thirdpartylibs.xml:2]: this needs to be up a dir level

/thirdpartylibs.xml

File: templates/flamegraph.mustache [hunk 7]
```diff
@@ -172,8 +172,8 @@
 <div id="details" style="min-height: 1.5rem; clear: both;">
 </div>
 
-<script type="text/javascript" src="https://d3js.org/d3.v4.min.js"></script>
-<script type="text/javascript" src="https://cdn.jsdelivr.net/npm/d3-flame-graph@4.0.6/dist/d3-flamegraph.min.js"></script>
+<script type="text/javascript" src="/admin/tool/excimer/lib/d3/d3.min.js"></script>
```
Comment (brendanheywood) [inline templates/flamegraph.mustache:175]: It's better to route the serving of these through /theme/javascript.php which will add all the proper cache busting params and caching headers to it gets cached by cdn / varnish

File: lib/d3-flame-graph/readme_moodle.txt (hunk 2 above)
Comment (brendanheywood) [reply lib/d3-flame-graph/readme_moodle.txt:15]: also worth having a step to fully remove the dir contents first so we aren't merging files across versions

File: lib/thirdpartylibs.xml (hunk 6 above)
Comment (brendanheywood) [reply lib/thirdpartylibs.xml:2]: To know if this is working, look at this admin page:

/admin/thirdpartylibs.php

File: templates/flamegraph.mustache (hunk 7 above)
Comment (jaypha) [reply templates/flamegraph.mustache:175]: How does this work?

**Pull request merged**